bayes_kit/ensemble.py

```diff
@@ -1,68 +1,154 @@
+
+        Raises:
+            ValueError: If `a` is provided and is not greater than or equal to 1, `walker`s is provided and not strictly positive and even, or if the `init` is provided and is not an `NDArray` of shape `walker` x `dimension`
+        """
+        # if not isinstance(model, LogDensityModel):
+        #     raise TypeError("model must follow the protocol LogDensityModel")
+        if not (a is None or isinstance(a, float) or isinstance(a, int)):
+            raise TypeError(f"a must be None, float, or int, found {a=}")
+        if not (num_walkers is None or isinstance(num_walkers, int)):
+            raise TypeError(f"num_walkers must be int, found {type(num_walkers)=}")
+        if not (init is None or isinstance(init, np.ndarray)):
+            raise TypeError("init must be None or NDArray")
+        if not (seed is None or isinstance(seed, int) or isinstance(seed, np.random.BitGenerator) or isinstance(seed, np.random.Generator)):
+            raise TypeError("seed must be None, int, np.random.BitGenerator, or np.random.Generator; found {type(seed)=}")
+        self._model = model
+        self._dim = self._model.dims()
+        if a != None and np.float64(a) <= 1:
+            raise ValueError(f"stretch bound must be greater than or equal to 1; found {a=}")
+        self._a = np.float64(a or 2.0)
+        self._sqrt_a = np.sqrt(self._a)
+        self._inv_sqrt_a = 1 / self._sqrt_a
+        if num_walkers is None:
+            self._num_walkers = 2 * self._dim
+        else:
+            if num_walkers < 2 or num_walkers % 2 != 0:
+                raise ValueError(f"number of walkers must be strictly positive, even integer; found {num_walkers=}")
+            self._num_walkers = num_walkers
+        self._half_num_walkers = self._num_walkers // 2
+        self._drawshape = (int(self._num_walkers), self._dim)
+        self._rng = np.random.default_rng(seed)
+        if init is None:
+            self._thetas = self._rng.normal(size=self._drawshape)
+        else:
+            if (init.shape != self._drawshape):
+                raise ValueError(f"init must be shape of draw {self._drawshape}; found {init.shape=}")
+            self._thetas = init
+        self._lp_thetas = [self._model.log_density(theta) for theta in self._thetas]
+        self._first_range = range(0, int(self._half_num_walkers))
+        self._second_range = range(int(self._half_num_walkers), int(self._num_walkers))
+
+    def __iter__(self) -> Iterator[Sample]:
+        """Return an infinite iterator for ensemble sampling.
+
+        Returns:
+            An iterator generating samples.
+        """
+        return self
+
+    def __next__(self) -> Sample:
+        """Return the next ensemble sample (`_num_walkers` x `_dim`).
+
+        Returns:
+            The next sample.
+        """
+        return self.sample()
+
+    def draw_z(self) -> Sample:
```

Comment:
- What's a `z`?

- Should we prefix "private" methods with an underscore, as we do with fields? Doing so can make it more obvious what's the intended API.

Reply:
`z` is the name of the variable in the paper.  

The doc explains how the function works:

```python
Return a random draw of `z` in `(1/a, a)` with `p(z) propto 1 / sqrt(z)`
```

The usage shows that it's an interpolation ratio:

```python
        z = self.draw_z()
        theta_star: NDArray[np.float64] = theta_j + z * (theta_k - theta_j)
```

Are you asking indirectly what it is because you think it should have a more descriptive variable name.

I'm OK prefixing internal methods with an underscore.  I wasn't even aware of that convention.  Please bear with me for a couple of years as I come up to speed on Python!